Metabase does not install new version of plugins when upgrading

**Describe the bug**
Metabase 0.34 included several time-related changes, but when upgrading Metabase it does not upgrade the plugin jars in the `plugins` directory resulting in errors when creating new databases. This will also cause queries on existing databases to break.

**Logs**
When trying to add a database for anything like Oracle, SQLite, or SQLServer, the first attempt at hitting "Save" results in this error in the browser:
```
Syntax error compiling at (metabase/driver/oracle.clj:1:1).
```

Subsequent attempts to save result in this error:
```
No method in multimethod 'connection-details->spec' for dispatch value: :oracle
```

Server logs for the first "Save" request:
```

01-13 08:42:00 INFO driver.impl :: Initializing driver :oracle...
01-13 08:42:00 INFO plugins.classloader :: Added URL file:/private/tmp/metabase-upgrade-plugins/plugins/oracle.metabase-driver.jar to classpath
01-13 08:42:00 DEBUG plugins.init-steps :: Loading plugin namespace metabase.driver.oracle...
01-13 08:42:00 ERROR driver.util :: Database connection error
Syntax error compiling at (metabase/driver/oracle.clj:1:1).
	at clojure.lang.Compiler.load(Compiler.java:7648)
	at clojure.lang.RT.loadResourceScript(RT.java:381)
	at clojure.lang.RT.loadResourceScript(RT.java:372)
	at clojure.lang.RT.load(RT.java:459)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.plugins.classloader$require.invokeStatic(classloader.clj:101)
	at metabase.plugins.classloader$require.doInvoke(classloader.clj:89)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at metabase.plugins.init_steps$fn__67727.invokeStatic(init_steps.clj:22)
	at metabase.plugins.init_steps$fn__67727.invoke(init_steps.clj:20)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.plugins.init_steps$do_init_steps_BANG_.invokeStatic(init_steps.clj:32)
	at metabase.plugins.init_steps$do_init_steps_BANG_.invoke(init_steps.clj:27)
	at metabase.plugins.lazy_loaded_driver$make_initialize_BANG_$fn__67766.invoke(lazy_loaded_driver.clj:58)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.driver.impl$initialize_if_needed_BANG_.invokeStatic(impl.clj:179)
	at metabase.driver.impl$initialize_if_needed_BANG_.invoke(impl.clj:160)
	at metabase.driver$the_initialized_driver.invokeStatic(driver.clj:146)
	at metabase.driver$the_initialized_driver.invoke(driver.clj:142)
	at metabase.driver$dispatch_on_initialized_driver.invokeStatic(driver.clj:154)
	at metabase.driver$dispatch_on_initialized_driver.doInvoke(driver.clj:149)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at clojure.lang.MultiFn.invoke(MultiFn.java:233)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__21110.invoke(util.clj:32)
	at metabase.util$do_with_timeout$fn__13131.invoke(util.clj:302)
	at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.FileNotFoundException: Could not locate metabase/util/date__init.class, metabase/util/date.clj or metabase/util/date.cljc on classpath.
	at clojure.lang.RT.load(RT.java:462)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5989)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:703)
	at metabase.driver.oracle$eval192$loading__6721__auto____193.invoke(oracle.clj:1)
	at metabase.driver.oracle$eval192.invokeStatic(oracle.clj:1)
	at metabase.driver.oracle$eval192.invoke(oracle.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7177)
	at clojure.lang.Compiler.eval(Compiler.java:7166)
	at clojure.lang.Compiler.load(Compiler.java:7636)
	... 50 more
01-13 08:42:00 DEBUG middleware.log :: POST /api/database 400 50.8 ms (0 DB calls)
{:valid false, :dbname "Syntax error compiling at (metabase/driver/oracle.clj:1:1).", :message "Syntax error compiling at (metabase/driver/oracle.clj:1:1)."}
```

**To Reproduce**
Steps to reproduce the behavior:
1. Download an older version of Metabase and run it
2. Download a newer jar and run that in the same directory
3. Note that none of the dates on the jars in `plugins/` change
4. Try to add a database for Oracle, SQLite, SQLServer, etc
5. See the error message mentioned above

**Expected behavior**
Running a new version of Metabase should upgrade the jar files in the `plugins` directory

**Screenshots**
n/a

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.117 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.2+9",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "11.0.2",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.2+9",
    "os.name": "Mac OS X",
    "os.version": "10.14.6",
    "user.language": "en",
    "user.timezone": "America/Los_Angeles"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "oracle"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "run-mode": "prod",
    "version": {
      "date": "2019-12-19",
      "tag": "v0.34.0",
      "branch": "release-0.34.x",
      "hash": "1335215"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Critical - breaks basic functionality and will also break existing database connections. 

**Additional context**
n/a

--- a/src/metabase/config.clj
+++ b/src/metabase/config.clj
@@ -48,8 +48,10 @@
 
 
 ;; These are convenience functions for accessing config values that ensures a specific return type
-;; TODO - These names are bad. They should be something like  `int`, `boolean`, and `keyword`, respectively.
-;; See https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#dont-repeat-namespace-alias-in-function-names for discussion
+;;
+;; TODO - These names are bad. They should be something like `int`, `boolean`, and `keyword`, respectively. See
+;; https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#dont-repeat-namespace-alias-in-function-names
+;; for discussion
 (defn ^Integer config-int  "Fetch a configuration key and parse it as an integer." [k] (some-> k config-str Integer/parseInt))
 (defn ^Boolean config-bool "Fetch a configuration key and parse it as a boolean."  [k] (some-> k config-str Boolean/parseBoolean))
 (defn ^Keyword config-kw   "Fetch a configuration key and parse it as a keyword."  [k] (some-> k config-str keyword))
@@ -81,9 +83,11 @@
         (into {} (for [[k v] props]
                    [(keyword k) v]))))))
 
+;; TODO - Can we make this `^:const`, so we don't have to read the file at launch when running from the uberjar?
 (def mb-version-info
   "Information about the current version of Metabase.
-   This comes from `resources/version.properties` for prod builds and is fetched from `git` via the `./bin/version` script for dev.
+   This comes from `resources/version.properties` for prod builds and is fetched from `git` via the `./bin/version`
+  script for dev.
 
      mb-version-info -> {:tag: \"v0.11.1\", :hash: \"afdf863\", :branch: \"about_metabase\", :date: \"2015-10-05\"}"
   (or (if is-prod?

--- a/src/metabase/util/files.clj
+++ b/src/metabase/util/files.clj
@@ -13,7 +13,7 @@
   (:import java.io.FileNotFoundException
            java.net.URL
            [java.nio.file CopyOption Files FileSystem FileSystems LinkOption OpenOption Path StandardCopyOption]
-           [java.nio.file.attribute FileAttribute FileTime]
+           java.nio.file.attribute.FileAttribute
            java.util.Collections))
 
 ;;; --------------------------------------------------- Path Utils ---------------------------------------------------
@@ -68,19 +68,20 @@
 
 ;;; ------------------------------------------------- Copying Stuff --------------------------------------------------
 
-(defn- last-modified-time ^FileTime [^Path path]
-  (Files/getLastModifiedTime path (u/varargs LinkOption)))
+(defn- last-modified-timestamp ^java.time.Instant [^Path path]
+  (when (exists? path)
+    (.toInstant (Files/getLastModifiedTime path (u/varargs LinkOption)))))
 
 (defn- copy-file! [^Path source, ^Path dest]
   (when (or (not (exists? dest))
-            (pos? (.compareTo (last-modified-time source) (last-modified-time dest))))
+            (not= (last-modified-timestamp source) (last-modified-timestamp dest)))
     (u/profile (trs "Extract file {0} -> {1}" source dest)
       (Files/copy source dest (u/varargs CopyOption [StandardCopyOption/REPLACE_EXISTING
                                                      StandardCopyOption/COPY_ATTRIBUTES])))))
 
 (defn copy-files!
-  "Copy all files in `source-dir` to `dest-dir`. Overwrites existing files if last modified date is older than that of
-  the source file."
+  "Copy all files in `source-dir` to `dest-dir`. Overwrites existing files if last modified timestamp is not the same as
+  that of the source file — see #11699 for more context."
   [^Path source-dir, ^Path dest-dir]
   (doseq [^Path source (files-seq source-dir)
           :let [target (append-to-path dest-dir (str (.getFileName source)))]]
@@ -111,11 +112,14 @@
       (f (get-path (.getPath url))))))
 
 (defmacro with-open-path-to-resource
-  "Execute `body` with an Path to a resource file (i.e. a file in the project `resources/` directory), cleaning up when
-  finished.
+  "Execute `body` with an Path to a resource file or directory (i.e. a file in the project `resources/` directory, or
+  inside the uberjar), cleaning up when finished.
 
   Throws a FileNotFoundException if the resource does not exist; be sure to check with `io/resource` or similar before
-  calling this."
+  calling this.
+
+    (with-open-path-to-resouce [path \"modules\"]
+       ...)"
   [[path-binding resource-filename-str] & body]
   `(do-with-open-path-to-resource
     ~resource-filename-str

--- a/src/metabase/util/files.clj
+++ b/src/metabase/util/files.clj
@@ -75,7 +75,8 @@
   (when (or (not (exists? dest))
             (pos? (.compareTo (last-modified-time source) (last-modified-time dest))))
     (u/profile (trs "Extract file {0} -> {1}" source dest)
-      (Files/copy source dest (u/varargs CopyOption [StandardCopyOption/REPLACE_EXISTING])))))
+      (Files/copy source dest (u/varargs CopyOption [StandardCopyOption/REPLACE_EXISTING
+                                                     StandardCopyOption/COPY_ATTRIBUTES])))))
 
 (defn copy-files!
   "Copy all files in `source-dir` to `dest-dir`. Overwrites existing files if last modified date is older than that of

